README.md
[![GoDoc](https://godoc.org/github.com/mg-castel/nagflux?status.svg)](https://godoc.org/github.com/mg-castel/nagflux)
[![Go Report Card](http://goreportcard.com/badge/Griesbacher/nagflux)](http:/goreportcard.com/report/Griesbacher/nagflux)
[![Circle CI](https://circleci.com/gh/Griesbacher/nagflux/tree/master.svg?style=svg)](https://circleci.com/gh/Griesbacher/nagflux/tree/master)
[![Coverage Status](https://coveralls.io/repos/Griesbacher/nagflux/badge.svg?branch=master&service=github)](https://coveralls.io/github/Griesbacher/nagflux?branch=master)
# Nagflux
#### A connector which transforms performancedata from Nagios/Icinga(2)/Naemon to InfluxDB/Elasticsearch
Nagflux collects data from the NagiosSpoolfileFolder and adds informations from Livestatus. This data is sent to an InfluxDB, to get displayed by Grafana. Therefor is the tool [Histou](https://github.com/mg-castel/histou) gives you the possibility to add Templates to Grafana.
<p>Nagflux can be seen as the process_perfdata.pl script from PNP4Nagios.</p>

## Dependencies

```
Golang 1.5+
```

## Install
```
go get -u github.com/mg-castel/nagflux
go build github.com/mg-castel/nagflux
```
A x86-64 Linux binary will be added to the releases. Here the link to the latest [Release](https://github.com/mg-castel/nagflux/releases/latest).

## Configure
Here are some of the important config-options:

| Section       | Config-Key    | Meaning       |
| ------------- | ------------- | ------------- |
|main|NagiosSpoolfileFolder|This is the folder where nagios/icinga writes its spoolfiles. Icinga2: `/var/spool/icinga2/perfdata`|
|main|NagfluxSpoolfileFolder|In this folder you can dump files with InfluxDBs linequery syntax, the will be shipped to the InfluxDB, the timestamp has to be in ms|
|main|FieldSeperator|This char is used to separate the logical parts of the tablenames. This char has to be an char which is not allowed in one of those: host-, servicename, command, perfdata|
|main|FileBufferSize|This is the size of the buffer which is used to read files from disk, if you have huge checks or a lot of them you maybe recive error messages that your buffer is too small and that's the point to change it|
|Log|MinSeverity|INFO is default an enough for the most. DEBUG give you a lot more data but it's mostly just spamming|
|InfluxDBGlobal|Version|Currentliy the only supported Version of InfluxDB is 0.9+|
|Influx "name"|Address|The URL of the InfluxDB-API|
|Influx "name"|Arguments|Here you can set your user name and password as well as the database. **The precision has to be ms!**|
|Influx "name"|NastyString/NastyStringToReplace|These keys are to avoid a bug in InfluxDB and should disappear when the bug is fixed|
|Influx "name"|StopPullingDataIfDown|This is used to tell Nagflux, if this Influxdb is down to stop reading new data. That's useful if you're using spoolfiles. But if you're using gearman set this always to false because by default gearman will not buffer the data endlessly|

## Start
If the configfile is in the same folder as the executable:
```
./nagflux
```
else:
```
./nagflux -configPath=/path/to/config.gcfg
```

## Debugging
- If the InfluxDB is not available Nagflux will stop and an log entry will be written.
- If the Livestatus is not available Nagflux will just write an log entry, but additional informations can't be gathered.
- If any part of the Tablename is not valid for the InfluxDB an log entry will written and the data is writen to a file which has the same name as the logfile just with the ending '.dump-errors'. You could fix the errors by hand and copy the lines in the NagfluxSpoolfileFolder
- If the Data can't be send to the InfluxDB, Nagflux will also write them in the '.dump-errors' file, you can handle them the same way.
- If the logs are showing files are being read (in DEBUG mode) but nothing is going into InfluxDB, check the perfdata template to ensure it matches OMD format. See [Perfdata Template](https://github.com/mg-castel/nagflux#perfdata-template) for more details.

## Dataflow
There are basically two ways for Nagflux to receive data:
- Spoolfiles: They are for useful if Nagflux is running at the same machine as Nagios
- Gearman: If you have a distributed setup, that's the way to go
<p>With both ways you could enrich your performance data with additional informations from livestatus. Like downtimes, notifications and so.<p>

Targets can be:

- **InfluxDB**, that's the main target and the reason for this project.
- Elasticsearch, more a prove of concept but it worked some time ago ;)
- JSON, to parse the data by an third tool. 

![Dataflow Image](https://raw.githubusercontent.com/Griesbacher/nagflux/master/doc/NagfluxDataflow.png "Nagflux Dataflow")

## OMD
Nagflux is fully integrated in [OMD-Labs](https://github.com/ConSol/omd), as well as Histou is. Therefor if you wanna try it out, it's maybe easier to install OMD-Labs.

## Perfdata Template
Nagflux supports a couple of Perfdata templates (see `main_test.go` for some supported formats). By default it assumes you have the [OMD formattemplate](https://github.com/ConSol/omd/blob/labs/packages/nagflux/skel/etc/nagflux/nagios_nagflux.cfg). If you are setting this up manually (not using OMD) please ensure your perfdata template is as follows:

**Host**
```
DATATYPE::HOSTPERFDATA\tTIMET::$TIMET$\tHOSTNAME::$HOSTNAME$\tHOSTPERFDATA::$HOSTPERFDATA$\tHOSTCHECKCOMMAND::$HOSTCHECKCOMMAND$
```

**Service**
```
DATATYPE::SERVICEPERFDATA\tTIMET::$TIMET$\tHOSTNAME::$HOSTNAME$\tSERVICEDESC::$SERVICEDESC$\tSERVICEPERFDATA::$SERVICEPERFDATA$\tSERVICECHECKCOMMAND::$SERVICECHECKCOMMAND$
```

If you are using Nagios the default templates will not work. Use the above templates with config `host_perfdata_file_template` and `service_perfdata_file_template`, respectively.

## DEMO
This Dockercontainer contains OMD and everything is preconfigured to use Nagflux/Histou/Grafana/InfluxDB: https://github.com/mg-castel/docker-omd-grafana

## Presentations
- Here is a presentation I held about Nagflux and Histou in 2016, only in German, sorry: [Slides](http://www.slideshare.net/PhilipGriesbacher/monitoring-workshop-kiel-2016-performancedaten-visualisierung-mit-grafana-influxdb)
- That's the first one from 2015, also only in German. [Slides](https://www.netways.de/fileadmin/images/Events_Trainings/Events/OSMC/2015/Slides_2015/Grafana_meets_Monitoring_Vorstellung_einer_Komplettloesung-Philip_Griesbacher.pdf) - [Video](https://www.youtube.com/watch?v=rY6N2H0UCFQ)


CHANGELOG.md
## v0.4.1 - 07.06.2017
### Feature
- Unicode support in Units like µs
- If performancedata contains only an U the value field will be missing, but a field unknown with the value true will be added

## v0.4.0 - 17.03.2017
### Feature
- Livestatusversion can be set in the config.
- JSONFileExport, to write collected data to an JSON File.

## v0.4.0-b1 - 16.03.2017
### Feature
- It is possible to define multiple Influxdbs, each addressed by an name, as well es Elasticsearchs, they are all called "targets". 
If the data contains an NAGFLUX:TARGET field, this one is used to direct this certain data to a specific target. 
If this tag is not added to the data, the config defines the default tag, by default "all" which means that the data is 
will be send to all targets. So "all" is a magic word and should not be uses as target name. See issue [#25](https://github.com/mg-castel/nagflux/issues/25).
- If an Influxdb is not reachable the option "StopPullingDataIfDown" decides if reading new data into Nagflux should go on or not.

### Fix
- Some minor bugs should be fixed.

## v0.3.1 - 13.03.2017
### Fix
- Allow comma separated Performancedata
- Spoolfilebuffer is now configurable
- Nagflux Tags and Fields are ignoring empty or malformed data. Valid but ignored Performancedate would be: NAGFLUX:TAG::$_SERVICENAGFLUX_TAG$ or NAGFLUX:TAG::  

## v0.3.0 - 07.12.2016
### Fix
- Duplicate values over perflabels on one service should be fixed

## v0.2.9 - 30.11.2016
### Feature
- If an Spoolfileline contains e.g. NAGFLUX:TAG::serv=server1 ID=1	NAGFLUX:FIELD::counter_a=123 counter_b=456 the serv and ID will be stored as tag and both counters as fields. This makes it possible to add addition information to your Nagiospoolfiles.

## v0.2.8 - 23.11.2016
### Fix
- out of memory error on big files

### Feature
- Livestatustimout is configurable
- Multiple ModGearman server are supported

## v0.2.7 - 16.11.2016
### Fix
- Livestatus index out of bound error
- Less Livestatus log entries on an error(downtime)

## v0.2.6 - 26.10.2016
### Feature
- Version is shown within the help message
- check_multi prefixes will be expanded if not done by the core

## v0.2.5 - 22.09.2016
### Fix
- Deadlock when InfluxDB is not running, again !?

## v0.2.4 - 20.09.2016
### Fix
- Deadlock when InfluxDB is not running
- Pass connection args when checking for database
- Missing logfile fix

### Feature
- use of vendor-folder
- Prometheus api

### Breaks
- When using go1.5 the envvar GO15VENDOREXPERIMENT should be set to 1 

## v0.2.3 - 08.09.2016
### Fix
- ignore selfsigned ssl certs
- livestatus detection improved
- wait for influxdb on start
- pause fileparsing when influxdb is not reachable
- skip non digit perfdata(U are ignored)


## v0.2.2 - 17.05.2016
### Fix
- livestatus ServiceNotifications with just 9 entries
- nagflux fileimport exception when column name is too short
- nagios livestatus query for performance issues

## v0.2.1.1 - 08.04.2016
### Features
- mod_gearman key will be cut if it's too long

## v0.2.1 - 30.03.2016
### Features
- mod_gearman support (experimental)

### Breaks
- New Nagflux import format

## v0.2.0 - 10.03.2016

### Features
- Elasticsearch support
- New Importformat

### Fixes
-  Version Bug

### Breaks
- The old InfluxDB layout is not valid anymore. To convert the old data use the Pythonscript in CONVERTER.

## v0.1.0 - 01.12.2015

### Features
- Everything :wink:


main_test.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/mg-castel/nagflux/target/influx"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"reflect"
	"testing"
	"time"
)

const (
	filename      = "test.gcfg"
	envInflux     = "NAGFLUX_TEST_INFLUX"
	envLivestatus = "NAGFLUX_TEST_LIVESTATUS"
	envSave       = "NAGFLUX_TEST_SAVE"
	databaseName  = "NAGFLUX_CI_TEST"
	timeout       = time.Duration(30) * time.Second
)

type testData struct {
	input  string
	output influx.SeriesValue
}

func testResult(t []testData, result []influx.SeriesValue) bool {
	hits := 0
	for _, testData := range t {
		for _, values := range result {
			if reflect.DeepEqual(values, testData.output) {
				hits++
				break
			}
		}
	}

	if hits == len(t) {
		return true
	}

	for _, testData := range t {
		found := false
		for _, values := range result {
			if reflect.DeepEqual(values, testData.output) {
				found = true
				break
			}
		}
		if !found {
			fmt.Fprintf(os.Stderr, "%s", fmt.Sprintln(testData.output))
			for _, values := range result {
				fmt.Fprintf(os.Stderr, "-- %s", fmt.Sprintln(values))
			}
		}
	}
	return false
}

var NagiosTestData1 = []testData{
	//Nasty
	{`DATATYPE::SERVICEPERFDATA	TIMET::1	HOSTNAME::h1	SERVICEDESC::s1	SERVICEPERFDATA::C: use=1;2;3;4;5	SERVICECHECKCOMMAND::usage
`,
		// [time command crit crit-fill host max min performanceLabel service value warn warn-fill]
		// [1000 usage 3 none h1 5 <nil> 4 C: use s1 <nil> 1 2 none]
		[]interface{}{1000.0, "usage", 3.0, "none", "h1", 5.0, nil, 4.0, `C: use`, "s1", nil, 1.0, 2.0, "none"}},
	{`DATATYPE::SERVICEPERFDATA	TIMET::3	HOSTNAME::h1	SERVICEDESC::s1	SERVICEPERFDATA::D:\ use=1;2;3;4;5	SERVICECHECKCOMMAND::usage
`,
		// [time command crit crit-fill host max min performanceLabel service value warn warn-fill]
		// [3000 usage 3 none h1 5 <nil> 4 D:\ use s1 <nil> 1 2 none]
		[]interface{}{3000.0, "usage", 3.0, "none", "h1", 5.0, nil, 4.0, `D:\ use`, "s1", nil, 1.0, 2.0, "none"}},
	//Normal
	{`DATATYPE::SERVICEPERFDATA	TIMET::2	HOSTNAME::h2	SERVICEDESC::s2	SERVICEPERFDATA::rta=2;3;4;5;6	SERVICECHECKCOMMAND::ping
`,
		//[2000 ping 4 none h2 6 <nil> 5 rta s2 <nil> 2 3 none]
		[]interface{}{2000.0, "ping", 4.0, "none", "h2", 6.0, nil, 5.0, "rta", "s2", nil, 2.0, 3.0, "none"}},
	{`DATATYPE::SERVICEPERFDATA	TIMET::1490957788	HOSTNAME::müü	SERVICEDESC::möö	SERVICEPERFDATA::getItinerary_min=34385µs	SERVICECHECKCOMMAND::check_perfs	SERVICESTATE::0	SERVICESTATETYPE::1
`,
		// [time command crit crit-fill host max min performanceLabel service value warn warn-fill]
		// [1.490957788e+12 check_perfs <nil> <nil> müü <nil> <nil> <nil> getItinerary_min möö µs 34385 <nil> <nil>]
		[]interface{}{1490957788000.0, "check_perfs", nil, nil, "müü", nil, nil, nil, "getItinerary_min", "möö", "µs", 34385.0, nil, nil}},
}
var NagiosTestData21 = []testData{
	//Database1
	{`DATATYPE::SERVICEPERFDATA	TIMET::4	HOSTNAME::h3	SERVICEDESC::s1	SERVICEPERFDATA::rta=2;3;4;5;6	SERVICECHECKCOMMAND::special	NAGFLUX:TARGET::` + database1 + `
`, //[2000 ping 4 none h2 6 <nil> 5 rta s2 2 3 none]
		[]interface{}{4000.0, "special", 4.0, "none", "h3", 6.0, nil, 5.0, "rta", "s1", nil, 2.0, 3.0, "none"}},
}
var NagiosTestData22 = []testData{
	//Database2
	{`DATATYPE::SERVICEPERFDATA	TIMET::4	HOSTNAME::h3	SERVICEDESC::s2	SERVICEPERFDATA::rta=2;3;4;5;6	SERVICECHECKCOMMAND::special	NAGFLUX:TARGET::` + database2 + `
`, //[2000 ping 4 none h2 6 <nil> 5 rta s2 2 3 none]
		[]interface{}{4000.0, "special", 4.0, "none", "h3", 6.0, nil, 5.0, "rta", "s2", nil, 2.0, 3.0, "none"}},
}

var NagfluxTestData1 = []testData{
	{`table&time&t_host&t_service&t_command&t_performanceLabel&f_value
metrics&10&nagflux&service1&command1&perf&20
`,
		//[10 command1 <nil> <nil> nagflux <nil> <nil> <nil> perf service1 20 <nil> <nil>]
		[]interface{}{10.0, "command1", nil, nil, "nagflux", nil, nil, nil, "perf", "service1", nil, 20.0, nil, nil}},
	{`metrics&20&nagflux&service\ 1&command1&perf\ 1&30
`,
		//[20 command1 <nil> <nil> nagflux <nil> <nil> <nil> perf\ 1 service\ 1 30 <nil> <nil>]
		[]interface{}{20.0, "command1", nil, nil, "nagflux", nil, nil, nil, "perf 1", "service 1", nil, 30.0, nil, nil}},
}

var NagfluxTestData2 = []testData{
	{`table&time&t_host&t_service&f_message
messages&100&nagflux&service1&"""Hallo World"""
`,
		//[100 <nil> <nil> <nil> nagflux <nil> Hallo World <nil> <nil> service1 <nil> <nil> <nil>]
		[]interface{}{100.0, nil, nil, nil, "nagflux", nil, "Hallo World", nil, nil, "service1", nil, nil, nil, nil}},
	{`messages&300&nagflux&service1&"""Hallo \\"""
`,
		//[300 <nil> <nil> <nil> nagflux <nil> Hallo \ <nil> <nil> service1 <nil> <nil> <nil>]
		[]interface{}{300.0, nil, nil, nil, "nagflux", nil, `Hallo \`, nil, nil, "service1", nil, nil, nil, nil}},
}

var TestDataName = `metrics`
var TestDataColumns = []string{"time", "command", "crit", "crit-fill", "host", "max", "message", "min", "performanceLabel", "service", "unit", "value", "warn", "warn-fill"}
var influxParam string
var livestatusParam string
var save bool
var finished chan bool

var database1 = databaseName + "_1"
var database2 = databaseName + "_2"

func init() {
	finished = make(chan bool)
	influxParam = os.Getenv(envInflux)
	if influxParam == "" {
		influxParam = "http://127.0.0.1:8086"
		fmt.Printf("%s is not set, using default: %s\n", envInflux, influxParam)
	}

	livestatusParam = os.Getenv(envLivestatus)
	if livestatusParam == "" {
		livestatusParam = "127.0.0.1:6557"
		fmt.Printf("%s is not set, using default: %s\n", envLivestatus, livestatusParam)
	}

	if os.Getenv(envSave) == "" {
		save = false
		fmt.Println("Will NOT save the database")
	} else {
		save = true
		fmt.Println("Will save the database")
	}
}

func TestEverything(t *testing.T) {
	os.Args = append(os.Args, "-configPath="+filename)
	go createTestData("test/nagios/", "1.txt", NagiosTestData1)
	go createTestData("test/nagios/", "2.txt", NagiosTestData21)
	go createTestData("test/nagios/", "3.txt", NagiosTestData22)
	go createTestData("test/nagflux/", "1.txt", NagfluxTestData1)
	go createTestData("test/nagflux/", "2.txt", NagfluxTestData2)
	createConfig()
	dropDatabase()
	go main()
	time.Sleep(time.Duration(1) * time.Second)
	os.Remove(filename)
	go checkDatabase()
	select {
	case <-finished:
	case <-time.After(timeout):
		result1, err1 := getEverything(database1)
		result2, err2 := getEverything(database2)
		dropDatabase()
		t.Errorf(
			"Expected data was not found in the influxdb within the timerange: %s\nError: %+v\nDatabase:%+v\nError: %+v\nDatabase:%+v",
			timeout, err1, result1, err2, result2,
		)
	}
	select {
	case quit <- true:
	case <-time.After(time.Duration(2) * time.Second):
		fmt.Println("Killing the test due to timeout")
	}
	dropDatabase()
}

func createTestData(folder, file string, data []testData) {
	if err := os.MkdirAll(folder, 0700); err != nil {
		panic(err)
	}
	fileData := []byte{}
	for _, data := range data {
		fileData = append(fileData, []byte(data.input)...)
	}
	if err := ioutil.WriteFile(folder+file, fileData, 0644); err != nil {
		panic(err)
	}
	fmt.Println(string(fileData))
}

func checkDatabase() {
	result := map[string]bool{
		"nagiosResult1":  false,
		"nagiosResult21": false,
		"nagiosResult22": false,
		"nagfluxResult1": false,
		"nagfluxResult2": false,
	}
	for {
		time.Sleep(time.Duration(500) * time.Millisecond)

		query, _ := getEverything(database1)
		if query == nil || len((*query).Results) == 0 {
			continue
		}
		result1 := (*query).Results[0]
		if len(result1.Series) != 2 || TestDataName != result1.Series[1].Name || !reflect.DeepEqual(TestDataColumns, result1.Series[1].Columns) {
			continue
		}

		query2, _ := getEverything(database2)
		if query2 == nil || len((*query2).Results) == 0 {
			continue
		}
		result2 := (*query2).Results[0]
		if len(result2.Series) != 2 || TestDataName != result2.Series[1].Name || !reflect.DeepEqual(TestDataColumns, result2.Series[1].Columns) {
			continue
		}
		if !result["nagiosResult1"] {
			result["nagiosResult1"] = testResult(NagiosTestData1, result1.Series[1].Values)
		}
		if !result["nagiosResult21"] {
			result["nagiosResult21"] = testResult(NagiosTestData21, result1.Series[1].Values)
		}
		if !result["nagiosResult22"] {
			result["nagiosResult22"] = testResult(NagiosTestData22, result2.Series[1].Values)
		}
		if !result["nagfluxResult1"] {
			result["nagfluxResult1"] = testResult(NagfluxTestData1, result1.Series[1].Values)
		}
		if !result["nagfluxResult2"] {
			result["nagfluxResult2"] = testResult(NagfluxTestData2, result1.Series[0].Values)
		}

		completed := true
		for k, v := range result {
			if !v {
				fmt.Println(k + " not forfilled")
				completed = false
			}
		}
		if completed {
			finished <- true
			return
		}
	}
}

func getEverything(database string) (*influx.ShowSeriesResult, error) {
	resp, err := http.Get(influxParam + "/query?db=" + url.QueryEscape(database) + "&q=select%20*%20from%20/.*/&epoch=ms&u=omdadmin&p=omd")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		var jsonResult influx.ShowSeriesResult
		json.Unmarshal(body, &jsonResult)
		return &jsonResult, nil
	}
	return nil, fmt.Errorf("Database query(%s) returned: %s", resp.Request.URL, resp.Status)
}

func dropDatabase() {
	if !save {
		http.Get(influxParam + "/query?u=omdadmin&p=omd&q=drop%20database%20" + url.QueryEscape(database1))
		http.Get(influxParam + "/query?u=omdadmin&p=omd&q=drop%20database%20" + url.QueryEscape(database2))
	}
}

func createConfig() {
	config := []byte(fmt.Sprintf(`
[main]
	NagiosSpoolfileFolder = "test/nagios"
	NagiosSpoolfileWorker = 1
	InfluxWorker = 2
	MaxInfluxWorker = 5
	DumpFile = "test/nagflux.dump"
	NagfluxSpoolfileFolder = "test/nagflux"
	FieldSeparator = "&"
	FileBufferSize = 65536
	BufferSize = 10000

[Log]
	LogFile = ""
	MinSeverity = "DEBUG"

[Monitoring]
	PrometheusAddress = ""

[InfluxDBGlobal]
	CreateDatabaseIfNotExists = true
	NastyString = ""
	NastyStringToReplace = ""
	HostcheckAlias = "hostcheck"

[InfluxDB "%s"]
	Enabled = true
	Version = 1.0
	Address = "%s"
	Arguments = "precision=ms&db=%s&u=omdadmin&p=omd"
	StopPullingDataIfDown = true

[InfluxDB "%s"]
	Enabled = true
	Version = 1.0
	Address = "%s"
	Arguments = "precision=ms&db=%s&u=omdadmin&p=omd"
	StopPullingDataIfDown = true

[InfluxDB "broken"]
	Enabled = true
	Version = 1.0
	Address = "http://127.0.0.1:1"
	Arguments = "precision=ms&db=broken&u=omdadmin&p=123456"
	StopPullingDataIfDown = false

[Livestatus]
	Type = "tcp"
	Address = "%s"
	MinutesToWait = 0

`, database1, influxParam, database1, database2, influxParam, database2, livestatusParam))
	if err := ioutil.WriteFile(filename, config, 0644); err != nil {
		panic(err)
	}
}


helper/influx_test.go
package helper

import (
	"github.com/mg-castel/nagflux/config"
	"reflect"
	"testing"
)

var SanitizeInfluxData = []struct {
	input  string
	output string
}{
	{"a a", `a\ a`},
	{"a,a", `a\,a`},
	{", ", `\,\ `},
	{"aa", "aa"},
	{`c:\ `, `c:\\ `},
	{"", ""},
	{`"a a"`, `"a a"`},
	{`§`, `SS`},
}

var SanitizeInfluxDataMap = []struct {
	input  map[string]string
	output map[string]string
}{
	{map[string]string{"a a": "'asdf'"}, map[string]string{`a\ a`: "asdf"}},
	{map[string]string{"": "a,a"}, map[string]string{"": `a\,a`}},
	{map[string]string{", ": "aa"}, map[string]string{`\,\ `: "aa"}},
	{map[string]string{`c:\ `: ""}, map[string]string{`c:\\ `: ""}},
	{map[string]string{"": ""}, map[string]string{"": ""}},
}

func TestSanitizeInfluxInput(t *testing.T) {
	//t.Parallel()
	config.InitConfigFromString(`[InfluxDBGlobal]
    # leave empty to disable
    NastyString = "§"
    NastyStringToReplace = "SS"
    HostcheckAlias = "hostcheck"
`)
	for _, data := range SanitizeInfluxData {
		actual := SanitizeInfluxInput(data.input)
		if actual != data.output {
			t.Errorf("SanitizeInfluxData(%s): expected: %s, actual: %s", data.input, data.output, actual)
		}
	}
}

func TestSanitizeMap(t *testing.T) {
	t.Parallel()
	for _, data := range SanitizeInfluxDataMap {
		actual := SanitizeMap(data.input)
		if !reflect.DeepEqual(actual, data.output) {
			t.Errorf("SanitizeInfluxData(%s): expected: %s, actual: %s", data.input, data.output, actual)
		}
	}
}


main.go
package main

import (
	"flag"
	"fmt"
	"github.com/mg-castel/nagflux/collector"
	"github.com/mg-castel/nagflux/collector/livestatus"
	"github.com/mg-castel/nagflux/collector/modGearman"
	"github.com/mg-castel/nagflux/collector/nagflux"
	"github.com/mg-castel/nagflux/collector/spoolfile"
	"github.com/mg-castel/nagflux/config"
	"github.com/mg-castel/nagflux/data"
	"github.com/mg-castel/nagflux/logging"
	"github.com/mg-castel/nagflux/statistics"
	"github.com/mg-castel/nagflux/target/elasticsearch"
	"github.com/mg-castel/nagflux/target/file/json"
	"github.com/mg-castel/nagflux/target/influx"
	"github.com/kdar/factorlog"
	"os"
	"os/signal"
	"syscall"
	"time"
)

//Stoppable represents every daemonlike struct which can be stopped
type Stoppable interface {
	Stop()
}

//Interval of the main loop, in which the amount of workers are calculated.
const updateRate = 1

//nagfluxVersion contains the current Github-Release
const nagfluxVersion string = "v0.4.1"

var log *factorlog.FactorLog
var quit = make(chan bool)

func main() {
	//Parse Args
	var configPath string
	var printver bool
	flag.Usage = func() {
		fmt.Println(`Nagflux by Philip Griesbacher`, nagfluxVersion, `
Commandline Parameter:
-configPath Path to the config file. If no file path is given the default is ./config.gcfg.
-V Print version and exit

For further informations / bugs reportes: https://github.com/mg-castel/nagflux
`)
	}
	flag.StringVar(&configPath, "configPath", "config.gcfg", "path to the config file")
	flag.BoolVar(&printver, "V", false, "print version and exit")
	flag.Parse()

	//Print version and exit
	if printver {
		fmt.Println(nagfluxVersion)
		os.Exit(0)
	}

	//Load config
	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		fmt.Printf("Can not find config file: '%s'.\n\nHelp:\n", configPath)
		flag.Usage()
		os.Exit(1)
	}
	config.InitConfig(configPath)
	cfg := config.GetConfig()

	//Create Logger
	logging.InitLogger(cfg.Log.LogFile, cfg.Log.MinSeverity)
	log = logging.GetLogger()
	log.Info(`Started Nagflux `, nagfluxVersion)
	log.Debugf("Using Config: %s", configPath)
	resultQueues := collector.ResultQueues{}
	stoppables := []Stoppable{}
	if len(cfg.Main.FieldSeparator) < 1 {
		panic("FieldSeparator is too short!")
	}
	pro := statistics.NewPrometheusServer(cfg.Monitoring.PrometheusAddress)
	pro.WatchResultQueueLength(resultQueues)
	fieldSeparator := []rune(cfg.Main.FieldSeparator)[0]

	for name, value := range cfg.InfluxDB {
		if value == nil || !(*value).Enabled {
			continue
		}
		influxConfig := (*value)
		target := data.Target{Name: name, Datatype: data.InfluxDB}
		config.StoreValue(target, false)
		resultQueues[target] = make(chan collector.Printable, cfg.Main.BufferSize)
		influx := influx.ConnectorFactory(
			resultQueues[target],
			influxConfig.Address, influxConfig.Arguments, cfg.Main.DumpFile, influxConfig.Version,
			cfg.Main.InfluxWorker, cfg.Main.MaxInfluxWorker, cfg.InfluxDBGlobal.CreateDatabaseIfNotExists,
			influxConfig.StopPullingDataIfDown, target, cfg.InfluxDBGlobal.ClientTimeout,
		)
		stoppables = append(stoppables, influx)
		influxDumpFileCollector := nagflux.NewDumpfileCollector(resultQueues[target], cfg.Main.DumpFile, target, cfg.Main.FileBufferSize)
		waitForDumpfileCollector(influxDumpFileCollector)
		stoppables = append(stoppables, influxDumpFileCollector)
	}

	for name, value := range cfg.Elasticsearch {
		if value == nil || !(*value).Enabled {
			continue
		}
		elasticConfig := (*value)
		target := data.Target{Name: name, Datatype: data.Elasticsearch}
		resultQueues[target] = make(chan collector.Printable, cfg.Main.BufferSize)
		config.StoreValue(target, false)
		elasticsearch := elasticsearch.ConnectorFactory(
			resultQueues[target],
			elasticConfig.Address, elasticConfig.Index, cfg.Main.DumpFile, elasticConfig.Version,
			cfg.Main.InfluxWorker, cfg.Main.MaxInfluxWorker, true,
		)
		stoppables = append(stoppables, elasticsearch)
		elasticDumpFileCollector := nagflux.NewDumpfileCollector(resultQueues[target], cfg.Main.DumpFile, target, cfg.Main.FileBufferSize)
		waitForDumpfileCollector(elasticDumpFileCollector)
		stoppables = append(stoppables, elasticDumpFileCollector)
	}

	for name, value := range cfg.JSONFileExport {
		if value == nil || !(*value).Enabled {
			continue
		}
		jsonFileConfig := (*value)
		target := data.Target{Name: name, Datatype: data.JSONFile}
		resultQueues[target] = make(chan collector.Printable, cfg.Main.BufferSize)
		templateFile := json.NewJSONFileWorker(
			log, jsonFileConfig.AutomaticFileRotation,
			resultQueues[target], target, jsonFileConfig.Path,
		)
		stoppables = append(stoppables, templateFile)
	}

	//Some time for the dumpfile to fill the queue
	time.Sleep(time.Duration(100) * time.Millisecond)

	liveconnector := &livestatus.Connector{log, cfg.Livestatus.Address, cfg.Livestatus.Type}
	livestatusCollector := livestatus.NewLivestatusCollector(resultQueues, liveconnector, cfg.Livestatus.Version)
	livestatusCache := livestatus.NewLivestatusCacheBuilder(liveconnector)

	for name, data := range cfg.ModGearman {
		if data == nil || !(*data).Enabled {
			continue
		}
		log.Infof("Mod_Gearman: %s - %s [%s]", name, (*data).Address, (*data).Queue)
		secret := modGearman.GetSecret((*data).Secret, (*data).SecretFile)
		for i := 0; i < (*data).Worker; i++ {
			gearmanWorker := modGearman.NewGearmanWorker((*data).Address,
				(*data).Queue,
				secret,
				resultQueues,
				livestatusCache,
			)
			stoppables = append(stoppables, gearmanWorker)
		}
	}

	log.Info("Nagios Spoolfile Folder: ", cfg.Main.NagiosSpoolfileFolder)
	nagiosCollector := spoolfile.NagiosSpoolfileCollectorFactory(
		cfg.Main.NagiosSpoolfileFolder,
		cfg.Main.NagiosSpoolfileWorker,
		resultQueues,
		livestatusCache,
		cfg.Main.FileBufferSize,
		collector.Filterable{Filter: cfg.Main.DefaultTarget},
	)

	log.Info("Nagflux Spoolfile Folder: ", cfg.Main.NagfluxSpoolfileFolder)
	nagfluxCollector := nagflux.NewNagfluxFileCollector(resultQueues, cfg.Main.NagfluxSpoolfileFolder, fieldSeparator)

	//Listen for Interrupts
	interruptChannel := make(chan os.Signal, 1)
	signal.Notify(interruptChannel, syscall.SIGINT)
	signal.Notify(interruptChannel, syscall.SIGTERM)
	go func() {
		<-interruptChannel
		log.Warn("Got Interrupted")
		stoppables = append(stoppables, []Stoppable{livestatusCollector, livestatusCache, nagiosCollector, nagfluxCollector}...)
		cleanUp(stoppables, resultQueues)
		quit <- true
	}()
	loop:
	//Main loop
	for {
		select {
		case <-time.After(time.Duration(updateRate) * time.Second):
		/*queriesSend, measureTime, err := statisticUser.GetData("send")
			if err != nil {
				continue
			}
			idleTime := (measureTime.Seconds() - queriesSend.Time.Seconds() / float64(influx.AmountWorkers())) / updateRate
			log.Debugf("Buffer len: %d - Idletime in percent: %0.2f ", len(resultQueues[0]), idleTime * 100)

		//TODO: fix worker spawn by type
			if idleTime > 0.25 {
				influx.RemoveWorker()
			} else if idleTime < 0.1 && float64(len(resultQueues[0])) > resultQueueLength * 0.8 {
				influx.AddWorker()
			}*/
		case <-quit:
			break loop
		}
	}
}

func waitForDumpfileCollector(dump *nagflux.DumpfileCollector) {
	if dump != nil {
		for i := 0; i < 30 && dump.IsRunning; i++ {
			time.Sleep(time.Duration(2) * time.Second)
		}
	}
}

//Wait till the Performance Data is sent.
func cleanUp(itemsToStop []Stoppable, resultQueues collector.ResultQueues) {
	log.Info("Cleaning up...")
	for i := len(itemsToStop) - 1; i >= 0; i-- {
		itemsToStop[i].Stop()
		time.Sleep(500 * time.Millisecond)
	}
	for _, q := range resultQueues {
		log.Debugf("Remaining queries %d", len(q))
	}
}
